Sources/main.cpp

```diff
@@ -318,27 +320,35 @@ static int GdbserverMain(int argc, char **argv) {
   std::string namedPipePath = opts.getString("named-pipe");
 
   bool reverse = opts.getBool("reverse-connect");
+  int fd = opts.getString("fd").empty()
+               ? -1
+               : std::atoi(opts.getString("fd").c_str());
 
   std::unique_ptr<Socket> socket;
-  if (opts.getPositional("[host]:port").empty()) {
-    // If we have a named pipe, set the port to 0 to indicate that we should
-    // dynamically allocate it and write it back to the FIFO.
-    socket = CreateTCPSocket(
-        gDefaultHost, namedPipePath.empty() ? gDefaultPort : "0", reverse);
+  if (fd >= 0) {
+    socket = ds2::make_unique<Socket>(fd);
   } else {
-    socket = CreateSocket(opts.getPositional("[host]:port"), reverse);
-  }
-
-  if (!namedPipePath.empty()) {
-    std::string realPort = socket->port();
-    FILE *namedPipe = fopen(namedPipePath.c_str(), "a");
-    if (namedPipe == nullptr) {
-      DS2LOG(Error, "unable to open %s: %s", namedPipePath.c_str(),
-             strerror(errno));
+    if (opts.getPositional("[host]:port").empty()) {
+      // If we have a named pipe, set the port to 0 to indicate that we should
+      // dynamically allocate it and write it back to the FIFO.
+      socket = CreateTCPSocket(
+          gDefaultHost, namedPipePath.empty() ? gDefaultPort : "0", reverse);
     } else {
-      // Write the null terminator to the file. This follows the llgs behavior.
-      fwrite(realPort.c_str(), 1, realPort.length() + 1, namedPipe);
-      fclose(namedPipe);
+      socket = CreateSocket(opts.getPositional("[host]:port"), reverse);
+    }
+
+    if (!namedPipePath.empty()) {
```

Comment:
If we pass the `--fd` argument, can we also use a named pipe? I'm kind of a noob when it comes to this part of ds2.

If we can combine these options, we can reformat the above code to be an `if, elif, else` block (creating a socket), and move the below block (using a named pipe) one scope out. Otherwise, this is fine.

Reply:
I think `--fd` will override `--named-pipe`. because we only use the named pipe to write the dynamically allocated port. but when fd is used, we don't even have a port. so I think fd should override host:port as well as named pipe

Reply:
@yuex is right, except I don't think `--fd` should necessarily override `--named-pipe`, but we probably want to display an error message instead.

Reply:
cool, I can add this